fs-index/src/index.rs

```diff
@@ -1,1120 +1,436 @@
+            id_to_resources
+                .entry(id)
+                .or_insert_with(Vec::new)
+                .push(resource.clone());
+        }
 
-            assert!(update.is_err())
-        })
+        let index = ResourceIndex {
+            root,
+            id_to_resources,
+            path_to_resource,
+        };
+        Ok(index)
     }
 
-    #[test]
-    fn update_one_should_index_delete_file_successfully() {
-        run_test_and_clean_up(|path| {
-            create_file_at(path.clone(), Some(FILE_SIZE_1), Some(FILE_NAME_1));
+    /// Update the index with the latest information from the file system
+    pub fn update_all(&mut self) -> Result<IndexUpdate<Id>> {
+        log::debug!("Updating index at root path: {:?}", self.root);
+        log::trace!("Current index: {:#?}", self);
+
+        let mut added = Vec::new();
+        let mut removed = Vec::new();
+
+        let current_paths = discover_paths(&self.root)?;
+
+        // Assuming that collection manipulation is faster than repeated
+        // lookups
+        let current_entries: HashMap<PathBuf, IndexedResource<Id>> =
+            scan_entries(current_paths.clone());
+        let previous_entries = self.path_to_resource.clone();
+        // `preserved_entries` is the intersection of current_entries and
+        // previous_entries
+        let preserved_entries: HashMap<PathBuf, IndexedResource<Id>> =
+            current_entries
+                .iter()
+                .filter_map(|(path, _resource)| {
+                    previous_entries.get(path).map(|prev_resource| {
+                        (path.clone(), prev_resource.clone())
+                    })
+                })
+                .collect();
 
-            let mut actual = ResourceIndex::build(path.clone());
+        // `created_entries` is the difference between current_entries and
+        // preserved_entries
+        let created_entries: HashMap<PathBuf, IndexedResource<Id>> =
+            current_entries
+                .iter()
+                .filter_map(|(path, resource)| {
+                    if preserved_entries.contains_key(path) {
+                        None
+                    } else {
+                        Some((path.clone(), resource.clone()))
+                    }
+                })
+                .collect();
 
-            let mut file_path = path.clone();
-            file_path.push(FILE_NAME_1);
-            std::fs::remove_file(file_path.clone())
-                .expect("Should remove file successfully");
+        // `updated_entries` is the difference between preserved_entries and
+        // current_entries where the last modified time has changed
+        // significantly
+        let updated_entries: HashMap<PathBuf, IndexedResource<Id>> =
+            preserved_entries
+                .iter()
+                .filter_map(|(path, resource)| {
+                    if current_entries.contains_key(path) {
+                        None
```

Comment:
`None` looks incorrect here.. We need a test case for this. In the original implementation, the condition was negated: `!curr.contains` what makes sense because updated paths are preserved and current.

And comment tells "difference", but should it be "intersection"?

Reply:
> And comment tells "difference", but should it be "intersection"?

Right. thanks for catching that.
Updated in d4d7e73